zebra-test/src/transcript.rs

```diff
@@ -1,59 +1,144 @@
 //! A [`Service`](tower::Service) implementation based on a fixed transcript.
 
-use color_eyre::eyre::{ensure, eyre, Report};
+use color_eyre::{
+    eyre::{eyre, Report, WrapErr},
+    section::Section,
+    section::SectionExt,
+};
 use futures::future::{ready, Ready};
 use std::{
     fmt::Debug,
     task::{Context, Poll},
 };
 use tower::{Service, ServiceExt};
 
-pub struct Transcript<R, S, I>
+type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+pub struct Transcript<R, S, E, I>
 where
-    I: Iterator<Item = (R, S)>,
+    I: Iterator<Item = (R, Result<S, E>)>,
 {
     messages: I,
 }
 
-impl<R, S, I> From<I> for Transcript<R, S, I>
+impl<R, S, E, I> From<I> for Transcript<R, S, E, I>
 where
-    I: Iterator<Item = (R, S)>,
+    I: Iterator<Item = (R, Result<S, E>)>,
 {
     fn from(messages: I) -> Self {
         Self { messages }
     }
 }
 
-impl<R, S, I> Transcript<R, S, I>
+impl<R, S, E, I> Transcript<R, S, E, I>
 where
-    I: Iterator<Item = (R, S)>,
+    I: Iterator<Item = (R, Result<S, E>)>,
     R: Debug,
     S: Debug + Eq,
+    E: Fn(Error) -> Result<(), Error>,
 {
     pub async fn check<C>(mut self, mut to_check: C) -> Result<(), Report>
     where
         C: Service<R, Response = S>,
-        C::Error: Debug,
+        C::Error: Into<Error>,
     {
         while let Some((req, expected_rsp)) = self.messages.next() {
             // These unwraps could propagate errors with the correct
             // bound on C::Error
-            let rsp = to_check.ready_and().await.unwrap().call(req).await.unwrap();
-            ensure!(
-                rsp == expected_rsp,
-                "Expected {:?}, got {:?}",
-                expected_rsp,
-                rsp
-            );
+            let fut = to_check.ready_and().await;
+
+            let fut = match fut {
+                Ok(fut) => Ok(fut),
+                Err(e) => match &expected_rsp {
+                    Ok(expected_rsp) => Err(eyre!(e.into()))
+                        .wrap_err("service's ready_and returned an error: transcript expected an response")
+                        .with_section(|| {
+                            format!("{:?}", expected_rsp).header("Expected Response:")
+                        }),
+                    Err(error_checker) => {
+                        error_checker(e.into())
+                            .map_err(|e| eyre!(e))
+                            .wrap_err("service returned an error but it didn't match the expected error")?;
+                        continue;
+                    },
+                },
+            }?;
```

Comment:
Because the transcript items are requests and responses to the service, I don't think it makes sense to check the result of poll_ready against the transcript.  We should require that the service does not fail during the execution of the transcript. Once a service returns an error from `poll_ready` it's assumed to be failed permanently (this is some possibly documented part of the `Service` contract), so there's not much information to be gained by testing failures of the service itself, rather than failures of individual requests.